lib/charms/oathkeeper/v0/forward_auth.py

```diff
@@ -0,0 +1,577 @@
+
+    auth_config_changed = EventSource(AuthConfigChangedEvent)
+    auth_config_removed = EventSource(AuthConfigRemovedEvent)
+
+
+class ForwardAuthRequirer(ForwardAuthRelation):
+    """Requirer side of the forward-auth relation."""
+
+    on = ForwardAuthRequirerEvents()
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        relation_name: str = RELATION_NAME,
+        forward_auth_requirer_config: Optional[RequirerConfig] = None,
+    ):
+        super().__init__(charm, relation_name)
+
+        self._charm = charm
+        self._relation_name = relation_name
+        self._forward_auth_requirer_config = forward_auth_requirer_config
+
+        events = self._charm.on[relation_name]
+        self.framework.observe(events.relation_created, self._on_relation_created_event)
+        self.framework.observe(events.relation_changed, self._on_relation_changed_event)
+        self.framework.observe(events.relation_departed, self._on_relation_departed_event)
+
+    def _on_relation_created_event(self, event: RelationCreatedEvent) -> None:
+        """Update the relation with requirer data when a relation is created."""
+        if not self.model.unit.is_leader():
+            return
+
+        self.update_requirer_relation_data(self._forward_auth_requirer_config, event.relation.id)
+
+    def _on_relation_changed_event(self, event: RelationChangedEvent) -> None:
+        """Get the forward-auth config and emit a custom config-changed event."""
+        if not self.model.unit.is_leader():
+            return
+
+        data = event.relation.data[event.app]
+        if not data:
+            logger.info("No provider relation data available.")
+            return
+
+        try:
+            forward_auth_data = _load_data(data, FORWARD_AUTH_PROVIDER_JSON_SCHEMA)
+        except DataValidationError as e:
+            logger.error(
+                f"Received invalid config from the provider: {e}. Config-changed will not be emitted."
+            )
+            return
+
+        decisions_address = forward_auth_data.get("decisions_address")
+        app_names = forward_auth_data.get("app_names")
+        headers = forward_auth_data.get("headers")
+
+        relation_id = event.relation.id
+        relation_app_name = event.relation.app.name
+
+        # Notify Traefik to update the routes
+        self.on.auth_config_changed.emit(
+            decisions_address, app_names, headers, relation_id, relation_app_name
+        )
+
+    def _on_relation_departed_event(self, event: RelationDepartedEvent) -> None:
+        """Notify the requirer that the relation has departed."""
+        self.on.auth_config_removed.emit(event.relation.id)
+
+    def update_requirer_relation_data(
+        self, ingress_app_names: Optional[RequirerConfig], relation_id: Optional[int] = None
+    ) -> None:
+        """Update the relation databag with app names that can get IAP protection."""
+        if not self.model.unit.is_leader():
+            return
```

Comment:
We have a `limit: 1` set so app data is somewhat equivalent to unit data, but I wonder if it makes more sense to use unit data for this. Perhaps it would help avoid some edge cases with leadership status?

Reply:
There should always be a leader which will write to the databag. What edge cases do you have in mind?

Reply:
Haven't examined all code paths but when we sprinkle leader guards we potentially need to make sure the logic is executed on [leader elected](https://discourse.charmhub.io/t/event-leader-elected/6470) too. Moreover, on [upgrade](https://discourse.charmhub.io/t/event-upgrade-charm/6485), we don't get relation events nor leader elected (only leader-settings-changed). So if after upgrade we need to update reldata, then we suddenly need to be mindful of that too.

If using unit data is possible, we won't need leader guards anymore.

Reply:
Aren't unit databags supposed to be used when each unit's data differs? In this case all units would write the same data. Having each unit write to its own databag could also introduce race conditions.
For edge cases, we can rely on `update-status` where the databag will be updated from `_process_status_and_configurations`.

Reply:
That's fair.
Not sure how the races would differ between app/unit data, but scaling traefik is kinda hypothetical.
As this is internal to the lib, could always change it later per need.